Add milestone labels

bl00mber opened the pull request:
close #17244

![gitea-milestone-labels2](https://user-images.githubusercontent.com/16987322/182328871-f3679594-b8a3-4d51-bdb9-e2267d9197e3.gif)

bl00mber commented:
Can you merge @lafriks @silverwind @6543 @techknowlogick @lunny

lunny commented on models/issues/milestone.go:
```diff
@@ -43,7 +43,9 @@ func (err ErrMilestoneNotExist) Error() string {
 type Milestone struct {
 	ID              int64                  `xorm:"pk autoincr"`
 	RepoID          int64                  `xorm:"INDEX"`
+	OwnerID         int64                  `xorm:"INDEX"`
```
Why we need this column?

lunny commented on models/issues/milestone.go:
```diff
@@ -118,6 +120,28 @@ func NewMilestone(m *Milestone) (err error) {
+		// During the session, SQLite3 driver cannot handle retrieve objects after update something.
+		// So we have to get all needed labels first.
+		labels := m.Labels
+		if err = sess.Find(&m.Labels); err != nil {
```
I don't think this is right. It will load all labels in the database.

lunny commented on models/issues/milestone.go:
```diff
@@ -177,6 +213,42 @@ func UpdateMilestone(m *Milestone, oldIsClosed bool) error {
 	}
 
+	var dbLabels []*Label
+	dbLabels, err = GetLabelsByMilestoneID(m.ID)
```
dbLabels, err :=

lunny replied:
ctx should be the first parameter

lunny commented on models/issues/milestone_label.go:
```diff
@@ -0,0 +1,305 @@
+}
+
+func (m *Milestone) clearLabels(ctx context.Context, e db.Engine, doer *user_model.User) (err error) {
+	if err = m.loadLabels(e); err != nil {
```
loadLabels' parameter should be changed to context.Context.

lunny commented on models/issues/milestone_label.go:
```diff
@@ -0,0 +1,305 @@
+	}
+	defer committer.Close()
+
+	if err = m.loadLabels(db.GetEngine(ctx)); err != nil {
```
As above.

lunny commented on models/issues/milestone_label.go:
```diff
@@ -0,0 +1,305 @@
+	toRemove = append(toRemove, m.Labels[removeIndex:]...)
+
+	if len(toAdd) > 0 {
+		if err = m.addLabels(ctx, db.GetEngine(ctx), toAdd, doer); err != nil {
```
As above.

lunny commented on models/migrations/migrations.go:
```diff
@@ -378,7 +378,7 @@ var migrations = []Migration{
 	NewMigration("v208 was completely broken - remigrate", remigrateU2FCredentials),
 
-	// Gitea 1.16.2 ends at v211
+	// Gitea 1.16.2 ends at v212
```
Why change this?

lunny commented:
Is it necessary to display milestone's lables in issue view interface? It may made some confusion.

6543 commented:
please resolve conflicts

bl00mber commented:
@6543 I will update today

bl00mber commented:
@lunny I will fix according to review

lunny commented on models/issues/label.go:
```diff
@@ -571,9 +571,9 @@ func updateLabelCols(ctx context.Context, l *Label, cols ...string) error {
 		SetExpr("num_milestones",
 			builder.Select("count(*)").From("milestone_label").
-				InnerJoin("milestone", "milestone_label.milestone_id = milestone.id").
+				InnerJoin("milestone", "milestone_label.label_id = milestone.id").
```
This change is not right.

bl00mber replied:
Retracted

lunny commented on models/issues/milestone.go:
```diff
@@ -123,6 +124,32 @@ func NewMilestone(m *Milestone) (err error) {
+	sess := db.GetEngine(db.DefaultContext)
+
+	if len(m.Labels) > 0 {
+		// During the session, SQLite3 driver cannot handle retrieve objects after update something.
```
Don't know what did you mean. I don't think we need to do that.

bl00mber replied:
This is the same as [here](https://github.com/go-gitea/gitea/blob/82ecd3b19eb1382521b7999bd0e2dd977c3b994d/models/issues/issue.go#L1022)

lunny commented on models/issues/milestone.go:
```diff
@@ -143,20 +170,32 @@ func GetMilestoneByRepoID(ctx context.Context, repoID, id int64) (*Milestone, er
 	} else if !has {
 		return nil, ErrMilestoneNotExist{ID: id, RepoID: repoID}
 	}
+	var labels []*Label
```
Labels should be loaded when necessary.

bl00mber replied:
Does it means they need to be loaded in [ListMilestoneLabels](https://github.com/go-gitea/gitea/blob/eaae336808cb80cf5628f540d0f877bc171458ac/routers/api/v1/repo/milestone_label.go#L17)?

lunny replied:
You can have a `LoadLabels` methods to lazyly load labels.

lunny commented on models/issues/milestone.go:
```diff
@@ -143,20 +170,32 @@ func GetMilestoneByRepoID(ctx context.Context, repoID, id int64) (*Milestone, er
 		return nil, ErrMilestoneNotExist{Name: name, RepoID: repoID}
 	}
-	return &mile, nil
+	var labels []*Label
```
As above

lunny commented on models/issues/milestone.go:
```diff
@@ -344,6 +424,7 @@ func (milestones MilestoneList) getMilestoneIDs() []int64 {
 type GetMilestonesOption struct {
 	db.ListOptions
 	RepoID   int64
+	Repo     *repo_model.Repository
```
I don't think we need this.

bl00mber replied:
It is needed in [toLabel](https://github.com/go-gitea/gitea/blob/eaae336808cb80cf5628f540d0f877bc171458ac/modules/convert/issue.go#L192), [ToAPIMilestone](https://github.com/go-gitea/gitea/blob/eaae336808cb80cf5628f540d0f877bc171458ac/modules/convert/issue.go#L222), [NewMilestoneLabels](https://github.com/go-gitea/gitea/blob/eaae336808cb80cf5628f540d0f877bc171458ac/models/issues/milestone_label.go#L73).
Should these values be made into separate properties of GetMilestonesOption?

bl00mber replied:
[Here](https://github.com/go-gitea/gitea/blob/eaae336808cb80cf5628f540d0f877bc171458ac/models/issues/label.go#L620), and in [issue](https://github.com/go-gitea/gitea/blob/eaae336808cb80cf5628f540d0f877bc171458ac/models/issues/issue.go#L111) for example repo is also added

lunny replied:
I mean `Repo` should be assigned out of the function.

bl00mber replied:
Can you please specify what it means to be assigned out of the function?

lunny replied:
Remove this new field of this struct and assign repo after invoking this function.

lunny commented on models/issues/milestone_label.go:
```diff
@@ -0,0 +1,299 @@
+}
+
+func hasMilestoneLabel(ctx context.Context, milestoneID, labelID int64) bool {
+	has, _ := db.GetEngine(ctx).Where("milestone_id = ? AND label_id = ?", milestoneID, labelID).Get(new(MilestoneLabel))
```
Use Exist and error should be handled.

bl00mber commented:
@lunny updated

lunny commented on models/issues/milestone.go:
```diff
@@ -326,6 +380,15 @@ func DeleteMilestoneByRepoID(repoID, id int64) error {
 		return err
 	}
+
+	if err = m.LoadLabels(); err != nil {
```
Need `ctx`

bl00mber replied:
Any idea why build failing? It says not enough arguments to db.TxContext(), but there are no arguments to be passed to db.TxContext()

bl00mber replied:
@lunny have you seen my question? I tried to make all calls with context but same error was produced

lunny commented on models/issues/milestone.go:
```diff
@@ -182,6 +198,44 @@ func UpdateMilestone(m *Milestone, oldIsClosed bool) error {
 		}
 	}
 
+	if err = m.LoadLabels(); err != nil {
```
Need `ctx` because it's in a transaction, otherwise it will deadlock in sqlite.

lunny commented on models/issues/milestone.go:
```diff
@@ -326,6 +380,15 @@ func DeleteMilestoneByRepoID(repoID, id int64) error {
 		return err
 	}
+
+	if err = m.LoadLabelsWithContext(ctx); err != nil {
```
So I think merge the 2 functions `LoadLabels` and `LoadLabelsWithContext` as one is a good choice.

lunny commented on models/issues/milestone_label.go:
```diff
@@ -0,0 +1,318 @@
+}
+
+// HasMilestoneLabel returns true if milestone has been labeled.
+func HasMilestoneLabel(milestoneID, labelID int64) (bool, error) {
```
Merge the two functions.

silverwind commented:
Were the review comments resolved? Also we'll need a rebase.

bl00mber commented:
@silverwind yes resolved & rebased

bl00mber commented:
@silverwind yes resolved & rebased

lunny commented on models/issues/milestone.go:
```diff
@@ -325,6 +379,15 @@ func DeleteMilestoneByRepoID(repoID, id int64) error {
 		return err
 	}
+
+	if err = m.LoadLabels(ctx); err != nil {
```
Maybe just exec a SQL to delete by milestone_id is better.

bl00mber replied:
@lunny how it should look like? There seems to be no db.Exec SQL queries that delete in this project

lunny commented:
You missed database migrations

wxiaoguang commented:
One more thing, there is a shared template for "label list", could the "milestone_new.tmpl" reuse the shared template?

And it looks strange to use `<form class="ui comment` (the `comment`) for a milestone form.

silverwind commented:
Please resolve all resolved conversations, so we know what's left to do.

bl00mber commented:
@silverwind all updated according to comments, except:
"SQL to delete" not added
"label list" shared template not used, because it is a different UI component that looks like table

wxiaoguang commented on routers/web/repo/milestone.go:
```diff
@@ -141,11 +205,40 @@ func NewMilestonePost(ctx *context.Context) {
+		if err != nil {
+			return
+		}
+		labelIDMark := Int64sToMap(labelIDs)
```
Could it use `container.Set[T]` directly?

wxiaoguang commented on templates/repo/issue/view_content/sidebar.tmpl:
```diff
@@ -166,6 +166,22 @@
 					</a>
+					{{if .Issue.Milestone.Labels}}
+						<div class="ui labels list">
+							<span class="labels-list">
```
IIRC, there is `RenderLabel` / `RenderLabels` in `helper.go`, other label pages use these functions?

lunny commented:
I still think both issue labels and milestone labels displayed in the issue sidebar are confusing.

wxiaoguang commented:
> I still think both issue labels and milestone labels displayed in the issue sidebar are confusing.

Hmm .... yup, so, let's first clarify the requirement before reviewing?

IMO the root question is: why milestones should have labels? If the labels are for mainly filtering purpose, does it make sense to attach labels to milestones?

silverwind commented:
Hmm I think if someone is really so heavily using milestones to needing labels on them, it's a somewhat valid use case and they can exist in milestone lists, but I still would not litter the issue/pr view with these second set of labels, it's too confusing.

bl00mber commented:
Update:
* used container.Set[T]
* deleted milestone labels from issue sidebar
* added milestone filtering by label

> it looks strange to use `<form class="ui comment` (the `comment`) for a milestone form.

This is the requirement for label select dropdown to work. You can see that [issue/new_form.tmpl](https://github.com/go-gitea/gitea/blob/main/templates/repo/issue/new_form.tmpl) also has the `comment` class.

6543 commented on models/issues/milestone_label.go:
```diff
@@ -0,0 +1,305 @@
+}
+
+func init() {
+	db.RegisterModel(new(MilestoneLabel))
```
on repo deletion, repo lables get deleted, we need to make sure that milestoneLabel that reference to them get deleted too

bl00mber replied:
Added deletion of MilestoneLabels

silverwind commented:
Will test soon.

silverwind commented on templates/repo/issue/milestone_new.tmpl:
```diff
@@ -39,6 +39,40 @@
 				</div>
+				<div class="ui segment metas">
+					<input id="label_ids" name="label_ids" type="hidden" value="{{.label_ids}}">
+					<div class="ui floating jump select-label dropdown">
```
```suggestion
					<div class="ui floating jump select-label dropdown upward">
```

Better to open upwards to not destroy page layout when opening the dropdown.

silverwind commented:
There is a bug in the text color rendering on the milestone list, the text is white:

<img width="441" alt="Screenshot 2023-05-01 at 01 00 26" src="https://user-images.githubusercontent.com/115237/235379964-52ce8b2d-a389-4564-95d4-88c6bd4a7d8e.png">

Color is correct on milestone edit page:

<img width="196" alt="Screenshot 2023-05-01 at 01 01 25" src="https://user-images.githubusercontent.com/115237/235379973-c9cdc4aa-c98f-4b31-a42e-671c496df90c.png">

Also correct in the issue list:

<img width="313" alt="Screenshot 2023-05-01 at 01 02 34" src="https://user-images.githubusercontent.com/115237/235380002-d2593812-9076-4ecf-9a98-4818aab6da6e.png">

silverwind commented on templates/repo/issue/milestones.tmpl:
```diff
@@ -107,6 +107,23 @@
+									<a
+										class="ui label item"
+										id="label_{{.ID}}"
+										style="color: white; background-color: {{.Color}}"
```
This static white here is incorrect. Check how other labels are rendered. The color should be derived dynamically from the background color.

wxiaoguang replied:
https://github.com/go-gitea/gitea/pull/20620#discussion_r1174495836

I guess `RenderLabel`  `RenderLabels` could handle these colors.

lunny commented:
How about remove milestone labels in issue page and keep milestone list/detail pages?

bl00mber commented:
> How about remove milestone labels in issue page and keep milestone list/detail pages?

They had been removed in previous update

lunny commented:
> > How about remove milestone labels in issue page and keep milestone list/detail pages?
> 
> They were removed in previous update

It's better to update the screenshots on the issue content.

silverwind commented:
Is it ready for review again?

bl00mber commented:
@silverwind yes I replaced text to RenderLabel and added deletion of MilestoneLabels when repo is deleted

silverwind commented:
Ok, I will test again later.

wxiaoguang commented on models/issues/milestone_label.go:
```diff
@@ -0,0 +1,318 @@
+}
+
+// NewMilestoneLabel creates a new milestone-label relation.
+func NewMilestoneLabel(m *Milestone, label *Label, doer *user_model.User) (err error) {
+	hasLabel, err := HasMilestoneLabel(db.DefaultContext, m.ID, label.ID)
+	if hasLabel {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	ctx, committer, err := db.TxContext(db.DefaultContext)
+	if err != nil {
+		return err
+	}
+	defer committer.Close()
```
It should use `ctx` argument, instead of starting a new transaction.

bl00mber replied:
There are both functions for each call, with context and without. There are more than 1200 db.DefaultContext uses in the project. Those that do not use context start with small letter. @silverwind @lunny should db.DefaultContext be deleted?

wxiaoguang commented on routers/web/repo/milestone.go:
```diff
@@ -174,6 +265,76 @@ func EditMilestone(ctx *context.Context) {
+		ctx.ServerError("GetLabelsByRepoID", err)
+		return
+	}
+	if len(selectLabels) > 0 && selectLabels != "0" {
+		selectLabelIDs, err = base.StringsToInt64s(strings.Split(selectLabels, ","))
+		if err != nil {
+			ctx.ServerError("StringsToInt64s", err)
+			return
+		}
+		// Get the exclusive scope for every label ID
+		labelExclusiveScopes = make([]string, 0, len(selectLabelIDs))
+		for _, labelID := range selectLabelIDs {
+			foundExclusiveScope := false
+			for _, label := range labels {
+				if label.ID == labelID || label.ID == -labelID {
+					labelExclusiveScopes = append(labelExclusiveScopes, label.ExclusiveScope())
+					foundExclusiveScope = true
+					break
+				}
+			}
+			if !foundExclusiveScope {
+				labelExclusiveScopes = append(labelExclusiveScopes, "")
+			}
+		}
+	}
+	hasSelected := len(selectLabelIDs) > 0
```
I guess I have seen such code not only once  😁 


could there be a shared function?

lunny commented:
Could we display labels on the milestone detail page.

silverwind commented:
> Could we display labels on the milestone detail page.

I agree but we shouldn't bloat the scope of this PR too much. Let's get it into a mergable state and refine later.

lunny commented:
> > Could we display labels on the milestone detail page.
> 
> I agree but we shouldn't bloat the scope of this PR too much. It's been waiting too long already. Let's get it into a mergable state and refine later.

I don't think so. Like issue labels will be displayed in issue list and issue detail page.
Currently milestone labels just are displayed in milestone list. They should also be displayed in milestone detail page.

bl00mber commented on models/issues/milestone.go:
```diff
@@ -374,6 +440,22 @@ func GetMilestones(opts GetMilestonesOption) (MilestoneList, int64, error) {
+		if err != nil {
+			log.Warn("Malformed Labels argument: %s", opts.Labels)
+		} else {
+			for i, labelID := range labelIDs {
+				if labelID > 0 {
+					sess.Join("INNER", fmt.Sprintf("milestone_label il%d", i),
+						fmt.Sprintf("milestone.id = il%[1]d.milestone_id AND il%[1]d.label_id = %[2]d", i, labelID))
+				} else {
+					sess.Where("milestone.id NOT IN (SELECT milestone_id FROM milestone_label WHERE milestone_id = ?)", -labelID)
+				}
+			}
```
This is same SQL as in [issue.go / getIssueStatsChunk](https://github.com/go-gitea/gitea/blob/d81659d03947ac4533099089c5f442437e1d6887/models/issues/issue.go#L1718)

bl00mber commented:
* labelExclusiveScopes taken out to a new function
* display labels on milestone detail page

I will update it only one more time, if there are any additional issues with this PR let me know for this update or I am closing it next time.

silverwind commented:
Bumped the migration number.

silverwind commented:
Tweaked labels in milestone list to look similar to issue list. The list has more problems, but not related to this PR.

<img width="831" alt="Screenshot 2023-05-21 at 22 57 41" src="https://github.com/go-gitea/gitea/assets/115237/8d43c4ab-f0d4-4e83-9d75-eae72b2a75f6">

silverwind reviewed (approved):
LGTM ui wise, can't speak for backend changes.

denyskon commented on routers/api/v1/repo/milestone_label.go:
```diff
@@ -0,0 +1,303 @@
+	ctx.JSON(http.StatusOK, convert.ToLabelList(selectLabels, ctx.Repo.Repository, ctx.Repo.Owner))
+}
+
+// ClearMilestoneLabels delete all the labels for a milestone
```
```suggestion
// ClearMilestoneLabels removes all labels from a milestone
```

denyskon commented:
@bl00mber Would be nice if we could merge this feature for 1.21. Could you update your branch and address the open review comments?

bl00mber commented:
@denyskon can you merge?

denyskon commented:
Thanks for the update, I'll review the PR soon

denyskon commented:
I'm sorry, I won't be able to review this large feature before the feature freeze. Unless any other maintainer is able to do so, I'll try to review this for 1.22.

yp05327 commented on models/issues/milestone_label.go:
```diff
@@ -0,0 +1,318 @@
+		return err
+	}
+
+	m.Labels = nil
+	if err = m.LoadLabels(ctx); err != nil {
+		return err
+	}
```
Are there any reasons that we need to clear it and load it from DB instead of adding new labels to `m.Labels` directly? 🤔
